### PR TITLE
[MKL-DNN] Batch norm mkl-dnn NHWC support

### PR DESCRIPTION
**Pull request by jczaja**
This PR implements NHWC Paddle layout as discussed in #20964 .

This PR is still under testing.

### CODE REVIEW

**Review by grygielski — APPROVED** [review]
Looks good to me. LGTM

File: paddle/fluid/operators/batch_norm_op.cc [hunk 1]
```diff
@@ -169,6 +170,32 @@ framework::OpKernelType BatchNormOp::GetExpectedKernelType(
...
+    const framework::OpKernelType &expected_kernel_type) const {
+#ifdef PADDLE_WITH_MKLDNN
+  // Only input require reshaping, weights and
+  // bias are having shape in NCHW order
+  if ((var_name == "X") &&
+      (expected_kernel_type.data_layout_ == framework::DataLayout::kMKLDNN) &&
+      (tensor.layout() != framework::DataLayout::kMKLDNN)) {
+    auto attrs = Attrs();
+    auto ar = paddle::framework::AttrReader(attrs);
+    const std::string data_layout = ar.Get<std::string>("data_layout");
```
Comment (LeoZhao-BJ) [inline paddle/fluid/operators/batch_norm_op.cc:184]: Why here use "data_layout" instead of "data_format", I saw in other PRs, here is data_format.

File: paddle/fluid/operators/batch_norm_op.cc [hunk 2]
```diff
@@ -465,6 +492,12 @@ framework::OpKernelType BatchNormGradOp::GetExpectedKernelType(
 #ifdef PADDLE_WITH_MKLDNN
   if (library == framework::LibraryType::kPlain &&
       platform::CanMKLDNNBeUsed(ctx)) {
+    // TODO(jczaja): Add support for NHWC
+    const std::string data_layout = ctx.Attr<std::string>("data_layout");
```
Comment (LeoZhao-BJ) [inline paddle/fluid/operators/batch_norm_op.cc:496]: same question here, why "data_layout"?

**Comment by LeoZhao-BJ** [comment]
I understand this attr name may be historic reason, but why not align, maybe @luotao1 can share some info ?

**Comment by luotao1** [comment]
It is indeed a historic reason, but I don't know the detail reason.

File: paddle/fluid/operators/batch_norm_op.cc (hunk 1 above)
Comment (jczaja) [reply paddle/fluid/operators/batch_norm_op.cc:184]: @LeoZhao-Intel Good observation. I have not chosen those names *data_format* and *data_layout*. Yes all other ops got *data_format* , but batch norm is having *data_layout* as described here: https://www.paddlepaddle.org.cn/documentation/docs/en/1.4/api/layers.html#batch-norm

I agree that names are inconsistent, but perhaps there was a reason after this design choice. 
Maybe @dzhwinter or @zhangting2020 can comment further on this names inconsistency.

**Comment by LeoZhao-BJ** [comment]
except attr name, LGTM.

**Comment by luotao1** [comment]
> This PR is still under testing

@jczaja Is the testing finished?

**Comment by jczaja** [comment]
@luotao1 Tests are still running. I will let you know today when it is done

**Comment by jczaja** [comment]
@luotao1 Ok. Verification of changes was completed. Please continue with review

**Review by luotao1 — APPROVED** [review]
LGTM
